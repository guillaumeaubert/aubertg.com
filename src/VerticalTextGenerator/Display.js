// @flow strict

import React from 'react';
import './Display.css';

function nest(text, i) {
  if (i === 0) {
    return [];
  }

  let nested = nest(text, i - 1);
  let rows = [text].concat([...text].slice(1));
  if (rows.length < 2) {
    rows[1] = ' ';
  }

  for (let j = 2; j < Math.max(nested.length + 2, rows.length); j++) {
    if (j >= rows.length) {
      rows[j] = ' ';
    }

    if (j < nested.length + 2) {
      rows[j] += '  ' + nested[j - 2];
    }
  }

  return rows;
}

type Props = {|
  +text: string,
  +transformation: string,
|};

class Display extends React.Component<Props> {
  render() {
    let {text, transformation} = this.props;

    let chars = [...text];
    let output = '';
    if (text.length > 0) {
      switch (transformation) {
        case 'vertical': {
          output = chars.join('\n');
          break;
        }
        case 'vtop': {
          output = text + '\n' + chars.slice(1).join('\n');
          break;
        }
        case 'vbottom': {
          let padding = ' '.repeat(chars.length - 1);
          output = chars
            .slice(0, -1)
            .map(x => padding + x)
            .join('\n');
          output += '\n' + text;
          break;
        }
        case 'plus': {
          let intersect = Math.floor((chars.length + 1) / 2);
          let padding = ' '.repeat(intersect - 1);
          let rows = chars.map(x => padding + x);
          rows[intersect - 1] = text;
          output = rows.join('\n');
          break;
        }
        case 'stairs': {
          for (let i = 0; i < 5; i++) {
            let padding = ' '.repeat(i * 3);
            output += padding + text + '\n\n';
          }
          break;
        }
        case 'slinky': {
          let reversed = chars.slice().reverse().join('');
          for (let i = 0; i < 5; i++) {
            let padding = ' '.repeat(i * (chars.length - 1));
            let vtext = chars
              .slice(i === 0 ? 0 : 1, -1)
              .map(x => padding + x)
              .join('\n');
            if (vtext.length > 0) {
              output += vtext + '\n';
            }
            output += padding + reversed + '\n';
          }
          break;
        }
        case 'nestedv': {
          output = nest(text, 5).join('\n');
          break;
        }
        case 'square': {
          if (text.length === 1) {
            output = text;
          } else {
            let padding = ' '.repeat(chars.length - 2);
            output = text + '\n';
            for (let i = 1; i < chars.length - 1; i++) {
              output += chars[i] + padding;
              output += chars.slice(-1 * i - 1, -1 * i).join('') + '\n';
            }
            output += chars.reverse().join('');
          }
          break;
        }
        case 'tree': {
          for (let i = chars.length; i > 0; i--) {
            output += ' '.repeat(chars.length - i)
              + Array(i).fill(chars[i - 1]).join(' ')
              + '\n';
          }
          break;
        }
        default: {
          output = '';
        }
      }
    }
    output = output.replace(/ /g, '\u00a0');
    output = output.split(/\n/).map(
      (row, index) => (
        <div key={`row_${index}`}>
          {row === '' ? '\u00a0' : row}
        </div>
      )
    );

    return (
      <div id="canvas">{output}</div>
    );
  }
}

export default Display;
