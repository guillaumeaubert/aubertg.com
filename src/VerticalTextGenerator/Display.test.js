// @flow strict

import React from 'react';
import Display from './Display';
import renderer from 'react-test-renderer';

it('The component renders correctly', () => {
  const tree = renderer.create(
    <Display text="DEVASTATING" transformation="square" />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
