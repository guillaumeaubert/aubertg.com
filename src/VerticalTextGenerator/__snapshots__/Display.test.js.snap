// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component renders "Nested Vs" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    DEVASTATING
  </div>
  <div>
    E
  </div>
  <div>
    V  DEVASTATING
  </div>
  <div>
    A  E
  </div>
  <div>
    S  V  DEVASTATING
  </div>
  <div>
    T  A  E
  </div>
  <div>
    A  S  V  DEVASTATING
  </div>
  <div>
    T  T  A  E
  </div>
  <div>
    I  A  S  V  DEVASTATING
  </div>
  <div>
    N  T  T  A  E
  </div>
  <div>
    G  I  A  S  V
  </div>
  <div>
       N  T  T  A
  </div>
  <div>
       G  I  A  S
  </div>
  <div>
          N  T  T
  </div>
  <div>
          G  I  A
  </div>
  <div>
             N  T
  </div>
  <div>
             G  I
  </div>
  <div>
                N
  </div>
  <div>
                G
  </div>
</div>
`;

exports[`The component renders "Plus" correctly 1`] = `
<div
  id="canvas"
>
  <div>
         D
  </div>
  <div>
         E
  </div>
  <div>
         V
  </div>
  <div>
         A
  </div>
  <div>
         S
  </div>
  <div>
    DEVASTATING
  </div>
  <div>
         A
  </div>
  <div>
         T
  </div>
  <div>
         I
  </div>
  <div>
         N
  </div>
  <div>
         G
  </div>
</div>
`;

exports[`The component renders "Slinky" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    D
  </div>
  <div>
    E
  </div>
  <div>
    V
  </div>
  <div>
    A
  </div>
  <div>
    S
  </div>
  <div>
    T
  </div>
  <div>
    A
  </div>
  <div>
    T
  </div>
  <div>
    I
  </div>
  <div>
    N
  </div>
  <div>
    GNITATSAVED
  </div>
  <div>
              E
  </div>
  <div>
              V
  </div>
  <div>
              A
  </div>
  <div>
              S
  </div>
  <div>
              T
  </div>
  <div>
              A
  </div>
  <div>
              T
  </div>
  <div>
              I
  </div>
  <div>
              N
  </div>
  <div>
              GNITATSAVED
  </div>
  <div>
                        E
  </div>
  <div>
                        V
  </div>
  <div>
                        A
  </div>
  <div>
                        S
  </div>
  <div>
                        T
  </div>
  <div>
                        A
  </div>
  <div>
                        T
  </div>
  <div>
                        I
  </div>
  <div>
                        N
  </div>
  <div>
                        GNITATSAVED
  </div>
  <div>
                                  E
  </div>
  <div>
                                  V
  </div>
  <div>
                                  A
  </div>
  <div>
                                  S
  </div>
  <div>
                                  T
  </div>
  <div>
                                  A
  </div>
  <div>
                                  T
  </div>
  <div>
                                  I
  </div>
  <div>
                                  N
  </div>
  <div>
                                  GNITATSAVED
  </div>
  <div>
                                            E
  </div>
  <div>
                                            V
  </div>
  <div>
                                            A
  </div>
  <div>
                                            S
  </div>
  <div>
                                            T
  </div>
  <div>
                                            A
  </div>
  <div>
                                            T
  </div>
  <div>
                                            I
  </div>
  <div>
                                            N
  </div>
  <div>
                                            GNITATSAVED
  </div>
  <div>
     
  </div>
</div>
`;

exports[`The component renders "Square" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    DEVASTATING
  </div>
  <div>
    E         N
  </div>
  <div>
    V         I
  </div>
  <div>
    A         T
  </div>
  <div>
    S         A
  </div>
  <div>
    T         T
  </div>
  <div>
    A         S
  </div>
  <div>
    T         A
  </div>
  <div>
    I         V
  </div>
  <div>
    N         E
  </div>
  <div>
    GNITATSAVED
  </div>
</div>
`;

exports[`The component renders "Stairs" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    DEVASTATING
  </div>
  <div>
     
  </div>
  <div>
       DEVASTATING
  </div>
  <div>
     
  </div>
  <div>
          DEVASTATING
  </div>
  <div>
     
  </div>
  <div>
             DEVASTATING
  </div>
  <div>
     
  </div>
  <div>
                DEVASTATING
  </div>
  <div>
     
  </div>
  <div>
     
  </div>
</div>
`;

exports[`The component renders "Tree" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    G G G G G G G G G G G
  </div>
  <div>
     N N N N N N N N N N
  </div>
  <div>
      I I I I I I I I I
  </div>
  <div>
       T T T T T T T T
  </div>
  <div>
        A A A A A A A
  </div>
  <div>
         T T T T T T
  </div>
  <div>
          S S S S S
  </div>
  <div>
           A A A A
  </div>
  <div>
            V V V
  </div>
  <div>
             E E
  </div>
  <div>
              D
  </div>
  <div>
     
  </div>
</div>
`;

exports[`The component renders "V, bottom" correctly 1`] = `
<div
  id="canvas"
>
  <div>
              D
  </div>
  <div>
              E
  </div>
  <div>
              V
  </div>
  <div>
              A
  </div>
  <div>
              S
  </div>
  <div>
              T
  </div>
  <div>
              A
  </div>
  <div>
              T
  </div>
  <div>
              I
  </div>
  <div>
              N
  </div>
  <div>
    DEVASTATING
  </div>
</div>
`;

exports[`The component renders "V, top" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    DEVASTATING
  </div>
  <div>
    E
  </div>
  <div>
    V
  </div>
  <div>
    A
  </div>
  <div>
    S
  </div>
  <div>
    T
  </div>
  <div>
    A
  </div>
  <div>
    T
  </div>
  <div>
    I
  </div>
  <div>
    N
  </div>
  <div>
    G
  </div>
</div>
`;

exports[`The component renders "Vertical" correctly 1`] = `
<div
  id="canvas"
>
  <div>
    D
  </div>
  <div>
    E
  </div>
  <div>
    V
  </div>
  <div>
    A
  </div>
  <div>
    S
  </div>
  <div>
    T
  </div>
  <div>
    A
  </div>
  <div>
    T
  </div>
  <div>
    I
  </div>
  <div>
    N
  </div>
  <div>
    G
  </div>
</div>
`;
